lib/dma.c

```diff
@@ -207,13 +215,7 @@ dma_controller_add_region(lm_ctx_t *lm_ctx, dma_controller_t *dma,
     region->size = size;
     region->page_size = page_size;
     region->offset = offset;
-
-    region->fd = dup(fd);       // dup the fd to get our own private copy
```

Comment:
This does not looks correct. In dma_controller_add_region() it holds reference on fd to call mmap() and in  dma_controller_remove_region() it munmap() and leaves the fd reference.
Now with this patch the fd reference is assumed to be held by caller and still the reference is left in   dma_controller_remove_region().
API wise this does not look correct. I think its better to keep these api's same and  the caller should take care of close(fd) if he has additional reference on the file.

Reply:
I agree with you, that's how I originally wanted to implement it but it turns out to be much more complicated in the kernel, so I've left it as a future optimization (https://github.com/nutanix/muser/issues/44). With this patch, the API changes as follows: `dma_controller_add_region`  operates with an fd as if it had opened it (it's installed by muser.ko) and it's responsible for closing it in the end.

Reply:
I do not see any reference to the above mentioned point in the issue (#44). Please update the issue accordingly, otherwise it will be forgotten.

Reply:
> I do not see any reference to the above mentioned point in the issue (#44). Please update the issue accordingly, otherwise it will be forgotten.

I've update the description in #44.